dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCResponseHeadersTest.java

```diff
@@ -0,0 +1,1490 @@
+*     IBM Corporation - initial API and implementation
+*******************************************************************************/
+package com.ibm.ws.fat.wc.tests;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.io.FileWriter;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import org.apache.hc.client5.http.classic.methods.HttpGet;
+import org.apache.hc.client5.http.classic.methods.HttpPost;
+import org.apache.hc.client5.http.entity.UrlEncodedFormEntity;
+import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
+import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
+import org.apache.hc.client5.http.impl.classic.HttpClientBuilder;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.NameValuePair;
+import org.apache.hc.core5.http.io.entity.EntityUtils;
+import org.apache.hc.core5.http.message.BasicNameValuePair;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.Collections;
+import java.util.List;
+
+import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.websphere.simplicity.config.HttpEndpoint;
+import com.ibm.websphere.simplicity.config.ServerConfiguration;
+import com.ibm.websphere.simplicity.log.Log;
+
+import componenttest.annotation.Server;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.ServerFileUtils;
+
+/**
+ * A set of tests to verify the <header> configurations and the corresponding response headers
+ */
+@RunWith(FATRunner.class)
+public class WCResponseHeadersTest {
+
+    private static final Class<?> ME = WCResponseHeadersTest.class;
+    private static final String APP_NAME = "ResponseHeadersTest";
+    private static final String APP_VERIFICATION_HEADER = "appVerificationHeader";
+    //App used to simulate responses during login process
+    private static final String APP_NAME_SECURE_APP = "SameSiteSecurityTest";
+
+    @Server("servlet40_headers")
+    public static LibertyServer server;
+
+    private static final ServerFileUtils serverFileUtils = new ServerFileUtils();
+
+   
+
+    // Tests can use this to indicate they don't make any config updates from the original configuration
+    private static boolean restoreSavedConfig = true; 
+
+    private static ServerConfiguration savedConfig; 
+
+    @Before
+    public void setUpBeforeEachTest() throws Exception {
+    
+        Exception failure = null;
+        String consoleLogFileName = WCResponseHeadersTest.class.getSimpleName() + ".log";
+        
+        if(!server.isStarted()){
+            server.updateServerConfiguration(savedConfig);
```

Comment:
Add check to ensure savedConfig isn't null? What happens if it is null?

Reply:
`savedConfig` is initialized by the `@BeforeClass `, where the `server` object is presumed to had already been initialized by the  runtime `@Server` annotation . I don't believe this should ever be null since it remains unchanged throughout the lifecycle of the test. If it were null, then each test would likely fail due to whatever configuration the server had on the previous test run. The sample documentation on the git wiki didn't suggest adding null checks on the `@BeforeClass` for the server, but it is a simple/quick check that. I'll add them, but I almost prefer the test fail and we find an exception that says that the setup failed, than it fail later on using a bad configuration and we end up having to debug to find that the issue wasn't in the test but on the configuration. 

It probably will be fairly obvious with how the tracing is setup, but some additional thoughts to consider.

Reply:
If you think you're better off without the null check I don't have any objection given the details you've provided.